docs/dev/topology-over-raft.md

```diff
@@ -273,6 +278,46 @@ migration track, then finalize can proceed and split each preexisting tablet int
 metadata. The replicas  will react to that by remapping its compaction groups into a new set which size
 is equal to the new tablet count.
 
+# Sharding with tablets
+
+Each table can have different shard assignment for a given token computed from the placement of tablet replicas,
+from table's tablet_map.
+
+Generic code should not use static sharders, which only work with vnode-based tables. So it should not use
+schema::get_sharder() or dht::static_shard_of(). It should use erm::get_sharder() instead:
+
+    table& t;
+    auto erm = t.erm();
+    dht::sharder& sharder = erm->get_sharder(); // valid as long as erm is alive
+
+A sharder obtained from effective_replication_map reflects the tablet_map in that particular version of topology.
+
+Since effective_replication_map_ptr blocks topology barriers, it should not be held for long. If the
+code is long-running but doesn't need to work with a particular topology version, it should use auto_refreshing_sharder.
+It is a sharder implementation which automatically switches to the latest effective_replication_map_ptr of the table when it changes.
+
+   dht::auto_refreshing_sharder sharder(table.shared_from_this());
+
```

Comment:
Need some words on the limitations of auto_refreshing_sharder (I imagine it's illegal to use information obtained from it across a preemption point)

Reply:
Done:

```
If you use auto_refreshing_sharder, the results of sharder methods may be invalid after preemption point,
since effective_replication_map instance used to obtain the results may no longer be alive. This
means that operations which use such a sharder may escape from topology barrier and not be waited for.
Such users should ensure that barriers synchronize with those operations in some other ways, for
example by using the topology guard mechanism.

```